File: src/main/java/io/sch/historyscan/infrastructure/features/codebase/list/ListCodeBaseManagement.java
package io.sch.historyscan.infrastructure.features.codebase.list;

import io.sch.historyscan.domain.contexts.codebase.find.CodeBasesListInventory;
import io.sch.historyscan.domain.contexts.codebase.find.CurrentCodeBase;
import io.sch.historyscan.infrastructure.features.codebase.info.CodeBaseInfoManagement;
import io.sch.historyscan.infrastructure.features.filesystem.FileSystemManager;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class ListCodeBaseManagement implements CodeBasesListInventory {

    private final String codebasesFolder;
    private final FileSystemManager fileSystemManager;
    private final CodeBaseInfoManagement codeBaseInfoManagement;

    public ListCodeBaseManagement(@Value("${io.sch.historyscan.codebases.folder}") String codebasesFolder,
                                  FileSystemManager fileSystemManager, CodeBaseInfoManagement codeBaseInfoManagement) {
        this.codebasesFolder = codebasesFolder;
        this.fileSystemManager = fileSystemManager;
        this.codeBaseInfoManagement = codeBaseInfoManagement;
    }

    @Override
    public List<CurrentCodeBase> listAll() {
        var folders = this.fileSystemManager.listFoldersFrom(codebasesFolder);
        return folders.stream()
                .map(codeBaseInfoManagement::codeBaseFromFolder)
                .toList();
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/infrastructure/features/codebase/list/ListCodeBaseManagement.kt
package io.sch.historyscan.infrastructure.features.codebase.list

import io.sch.historyscan.domain.contexts.codebase.find.CodeBasesListInventory
import io.sch.historyscan.domain.contexts.codebase.find.CurrentCodeBase
import io.sch.historyscan.infrastructure.common.filesystem.FileSystemManager
import io.sch.historyscan.infrastructure.features.codebase.info.CodeBaseInfoManagement
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
class ListCodeBaseManagement(
    @param:Value("\${io.sch.historyscan.codebases.folder}") private val codebasesFolder: String,
    private val fileSystemManager: FileSystemManager, private val codeBaseInfoManagement: CodeBaseInfoManagement
) : CodeBasesListInventory {
    override fun listAll(): List<CurrentCodeBase> {
        val folders = fileSystemManager.listFoldersFrom(codebasesFolder)
        return folders.map { folder ->
            folder.let { codeBaseInfoManagement.codeBaseFromFolder(folder) }
        }
    }
}
